cmd/gcp-controller-manager/dpwi/hms/hms.go

```diff
@@ -0,0 +1,176 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hms
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	apiserveroptions "k8s.io/apiserver/pkg/server/options"
+	"k8s.io/apiserver/pkg/util/webhook"
+	"k8s.io/client-go/rest"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+)
+
+const (
+	hmsRequestTimeout = 30 * time.Second
+)
+
+type HMSClient struct {
+	webhook *webhook.GenericWebhook
+}
+
+func NewHMSClient(url string, authProvider *clientcmdapi.AuthProviderConfig) (*HMSClient, error) {
+	config := &rest.Config{
+		Host:         url,
+		AuthProvider: authProvider,
+		Timeout:      hmsRequestTimeout,
+		QPS:          50,
+		Burst:        100,
+		ContentConfig: rest.ContentConfig{
+			NegotiatedSerializer: serializer.NegotiatedSerializerWrapper(runtime.SerializerInfo{}),
+		},
+	}
+	client, err := rest.UnversionedRESTClientFor(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create REST client for HMS from config %v: %v", config, err)
+	}
+	return &HMSClient{
+		webhook: &webhook.GenericWebhook{RestClient: client, RetryBackoff: *apiserveroptions.DefaultAuthWebhookRetryBackoff(), ShouldRetry: webhook.DefaultShouldRetry},
+	}, nil
+}
+
+// isErrorHTTPStatus returns true if statusCode is not in the range of 200 to 299 inclusive.
+func isErrorHTTPStatus(statusCode int) bool {
+	return statusCode < 200 || statusCode >= 300
+}
+
+func (h *HMSClient) Sync(node, zone string, gsaList []string) error {
+	sort.Strings(gsaList)
+	req := syncNodeRequest{
+		NodeName:  node,
+		NodeZone:  zone,
+		GSAEmails: gsaList,
+	}
+	return h.call(req, nil)
+}
+
+// Authorize implements the saMappingAuthorizer interface.  It calls HMS to verify if ksa has
+// permission to get certificates as gsa.
+func (h *HMSClient) Authorize(kns, ksa, gsa string) (bool, error) {
+	reqMapping := serviceAccountMapping{
+		KNSName:  kns,
+		KSAName:  ksa,
+		GSAEmail: gsa,
+	}
+	req := authorizeSAMappingRequest{
+		RequestedMappings: []serviceAccountMapping{reqMapping},
+	}
+
+	var rsp authorizeSAMappingResponse
+	if err := h.call(req, &rsp); err != nil {
+		return false, err
+	}
+
+	if permitted := rsp.PermittedMappings; len(permitted) > 0 && permitted[0] == reqMapping {
+		return true, nil
+	}
+	if denied := rsp.DeniedMappings; len(denied) > 0 && denied[0] == reqMapping {
+		return false, nil
+	}
+	return false, fmt.Errorf("internal error: requested mapping %v not found in response %+v", reqMapping, rsp)
+}
+
+func (h *HMSClient) call(req, rsp interface{}) error {
```

Comment:
Accept a context and add it to the post.

Reply:
Added, thank you!